Sprint 2/step 2

elrouss opened the pull request.

elrouss committed:
refactor & feat: init redux and file structure, rewrite adding ingredients in constructor with redux, add ingredients deletion from constructor

elrouss committed:
refactor: get ingredients from api with redux

elrouss committed:
refactor: save current ungredient data in redux store

elrouss committed:
refactor: create reducer for order details and delete useContext

elrouss committed:
Merge pull request #12 from elrouss/feat/redux-and-file-structure

INIT:
1) Install Redux Toolkit

REFACTOR:
1) Remove React useContext and useReducer and create Redux store, reducers, actions and dispatchers:
- All ingredients
- Selected ingredients (with new feature - deletion from burger constructor)
- Selected ingredient
- Order details

elrouss committed:
feat: create unique key for each selected ingredient, so multiple can be added or a single one deleted without console errors

elrouss committed:
feat: add drag&drop and plug for empty containers in burger constructor

elrouss committed:
refactor: create component burger-bun

elrouss committed:
feat: add dnd sorting of ingredients in constructor

elrouss committed:
Merge pull request #13 from elrouss/feat/dnd

REFACTOR:
1) Created 2 componens `BurgerBun` and `SelectedBurgerIngredients`;

FEAT:
1) Added plugs for empy containers in burger constructor;
2) Made dnd for putting ingredients in constructor and sorting them there;
3) Wrote new method in redux to sort ingredients in constructor

elrouss committed:
feat: add ingredients counter

elrouss committed:
Merge pull request #14 from elrouss/feat/ingredients-counter

FEAT:
1) Add ingredients counter

elrouss committed:
refactor: delete context and dispatcher in counting total price

elrouss committed:
Merge pull request #15 from elrouss/refactor/total-price

refactor: delete context and dispatcher in counting total price

elrouss committed:
feat: switch selected tab on scroll

elrouss committed:
feat: scroll in container on tab's click

elrouss committed:
Merge pull request #16 from elrouss/feat/tabs

FEAT:
1) Tabs selected changed on scroll in container;
2) Scroll in container on click

elrouss committed:
refactor: delete react contexts and reducers

elrouss committed:
feat: add preloader and error boundaries on api get data

elrouss committed:
feat: check PropTypes

elrouss committed:
refactor: rename folders in src

elrouss committed:
refactor: change register on new files with names

elrouss committed:
memoize burger ingredient on opening modal

elrouss committed:
refactor: change names of files

elrouss committed:
docs: update README

elrouss committed:
Merge pull request #17 from elrouss/refactor/final

FEAT:
1) Added preloader and error boundaries, while waiting data from server;

REFACTOR:
1) Changed letters' case in folders' and files' names;
2) Added PropTypes;
3) Memoized burger ingredient on opening modal

DOCS:
1) Updated README

elrouss committed:
docs: update README

kyzinatra commented on src/services/features/ingredients/reducer.jsx:
```diff
@@ -0,0 +1,15 @@
+import API from '../../../utils/constants';
+
+export const ingredientsApiReducer = createApi({
+  reducerPath: 'ingredientsApi',
```
Отлично! Круто, что вы пользуетесь toolkit

kyzinatra commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,132 @@
+    const TABS_BOTTOM_EDGE_VALUE =
+      tabsRef.current.getBoundingClientRect().bottom;
+    const BUNS_TOP_EDGE_VALUE = bunsRef.current.getBoundingClientRect().top;
+    const SAUCES_TOP_EDGE_VALUE = saucesRef.current.getBoundingClientRect().top;
```
Отлично! Хорошее именование констант

kyzinatra commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,132 @@
+
+  const handleModalOpen = useCallback((evt, ingredient) => {
+    if (evt.type === 'click' || evt?.key === 'Enter') {
+      dispatch(SHOW_INGREDIENT_DETAILS(ingredient));
```
Отлично! Очень круто, что вы сделали открытие модального окна на `enter`

kyzinatra commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,132 @@
+    if (isIngredientDetailsModalOpened) return;
+
+    // Time is the same as the animation of modals' appearing
+    setTimeout(() => dispatch(RESET_INGREDIENT_DETAILS()), 300);
```
Можно лучше: Время стоит вынести также в константу или, что было бы вообще очень здорово, использовать событие окончания анимации для вызова `RESET_INGREDIENT_DETAILS()`

kyzinatra commented on src/components/burger-ingredients/burger-ingredients-section/burger-ingredients-section.jsx:
```diff
@@ -0,0 +1,39 @@
+import styles from './burger-ingredients-section.module.scss';
+
+const BurgerIngredientsSection = forwardRef(
+  ({ typeRus, typeEng, ...rest }, ref) => {
```
Отлично! Правильное использование `forwardRef`

kyzinatra commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,237 @@
+
+// eslint-disable-next-line consistent-return
+async function sendOrder(order, dispatch) {
+  try {
```
Можно лучше: создание заказа лучше также перенести на `redux`. Для этого можно использовать, к примеру, `createAsyncThunk()`

kyzinatra commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,237 @@
+          {(selectedIngredients.length && (
+            <div className={styles.components}>
+              {selectedIngredients.map((ingredient, index) => (
+                <SelectedBurgerIngredient
```
Можно лучше: этот компонент кажется слишком сложным. Можно его разбить на несколько.

kyzinatra commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,237 @@
+    <>
+      <section aria-label="Оформление заказа">
+        <form className={styles.order} ref={drop}>
+          <BurgerBun
```
Можно лучше: форму стоит отправлять на `onSubmit={}`

kyzinatra commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,237 @@
+  // and putting them inside the constructor
+  const [{ isOver, ingredientTypeDrop }, drop] = useDrop(
+    () => ({
+      accept: DRAG_TYPES.INGREDIENT,
```
Отлично! Хорошая реализация DnD

kyzinatra commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,237 @@
+  // Selecting ingredients from left container
+  // and putting them inside the constructor
+  const [{ isOver, ingredientTypeDrop }, drop] = useDrop(
+    () => ({
```
Отлично! Круто, что вы комментируете код

kyzinatra commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,237 @@
+  const [totalPrice, setTotalPrice] = useState(0);
+  const [prevBunPrice, setPrevBunPrice] = useState(0);
+  const [isOrderDetailsModalOpened, setIsOrderDetailsModalOpened] =
+    useState(false);
```
Можно лучше: иногда стоит делать названия более короткими, если смысл не теряется `isModalOpened` (потому что в компоненте 1 модальное окно)

kyzinatra commented on src/components/burger-constructor/selected-burger-ingredient/selected-burger-ingredient.jsx:
```diff
@@ -0,0 +1,104 @@
+  return (
+    <div
+      className={`${styles.item}${
+        (isDragging && ` ${styles.itemDragging}`) || ''
```
Можно лучше:
Если создать функцию, которая бы реализовывала функционал:
```js
clx({
  Active: true,
  Active2: true,
  NotActive: false
}) // -> "Active Active2"
```
То можно будет удобно и коротко описывать условные классы. 

А для краткости записи без условий можно добавить интерфейс `clx([class1, class2])// -> "class1 class2"`

Тогда можно будет записать:

```js
clx({
  [styles.navlink]: true,
  [styles.active]: pathname === "/profile"
}) 
```

Это сократит код и сделает его более элегантным

kyzinatra commented:
Выглядит очень круто! Видно, что проделана большая работа. Но все же есть моменты на которые стоит обратить внимание.
Подробнее: 


**Можно лучше:**

-  Время стоит вынести также в константу или, что было бы вообще очень здорово, использовать событие окончания анимации для вызова `RESET_INGREDIENT_DETAILS()`
-  создание заказа лучше также перенести на redux. Для этого можно использовать, к примеру, `createAsyncThunk()`
-  этот компонент кажется слишком сложным. Можно его разбить на несколько.
-  форму стоит отправлять на `onSubmit={}`
-  иногда стоит делать названия более короткими, если смысл не теряется `isModalOpened` (потому что в компоненте 1 модальное окно)
- 
Если создать функцию, которая бы реализовывала функционал:

```js
clx({
  Active: true,
  Active2: true,
  NotActive: false
}) // -> "Active Active2"
```

То можно будет удобно и коротко описывать условные классы.

А для краткости записи без условий можно добавить интерфейс clx([class1, class2])// -> "class1 class2"

Тогда можно будет записать:

```js
clx({
  [styles.navlink]: true,
  [styles.active]: pathname === "/profile"
}) 
```

Это сократит код и сделает его более элегантным


**Что понравилось:**

-  Круто, что вы пользуетесь `toolkit`
-  Хорошее именование констант
-  Очень круто, что вы сделали открытие модального окна на enter
-  Правильное использование `forwardRef`
-  Хорошая реализация `DnD`
-  Круто, что вы комментируете код


**Обратите внимание, что работа принимается только после исправления всех `«Надо исправить»`.**

elrouss commented:
Данил, благодарю за проверку! Замечания учту и код рефакторну :)

elrouss committed:
refactor: add changes depending on review
- send order with createAsyncThunk (and rewrite method on createSlice)
- clear data in modal with animation end
- put submit on modal
